### PR TITLE
Add Tx Indexer Extension

### PR DESCRIPTION
**Pull request by aaronbuchwald**
This PR refactors the tx indexing from the MorpheusVM/TokenVM controller code into a separate module under `extensions/indexer`

This also fixes an apparent bug where the database created internally to the MorpheusVM/TokenVM for tx indexing would never be closed.

### CODE REVIEW

File: examples/tokenvm/tests/integration/integration_test.go
Comment (joshua-kim) [inline examples/tokenvm/tests/integration/integration_test.go:45]: Not going to block on this but we need to fix this repo's imports. There doesn't seem to be any rationale for how the imports are ordered currently so I keep having to manually undo the diffs my ide auto-organizes.

File: extensions/indexer/indexer.go
Comment (joshua-kim) [inline extensions/indexer/indexer.go:1]: nit: rename to `extension/indexer`. I think it's idiomatic in Go to use singular package names (except for when it doesn't, like `errors`, or `strings` which I believe is because it would collide with the existing primitive type).

File: extensions/indexer/indexer.go
Comment (joshua-kim) [inline extensions/indexer/indexer.go:24]: If we're going to make this an extension that other vms will pull in it probably makes sense to add unit tests for this

File: examples/morpheusvm/controller/action_indexer.go
Comment (aaronbuchwald) [reply examples/morpheusvm/controller/action_indexer.go:16]: We could rename it as well. We could more accurately call it an `AcceptedSubscriber`.

This would be more accurate to exactly what it's doing and an indexer would be one thing you could build as an `AcceptedSubscriber`. Similarly, you could pass in an `acceptedActionHandler` or similar, which would be just this for the MorpheusVM and would also maintain the in-memory order book in the case of TokenVM.

File: examples/morpheusvm/controller/controller.go
Comment (aaronbuchwald) [reply examples/morpheusvm/controller/controller.go:104]: This is just keeping the same pattern that was there before. The for loop currently does two things, it iterates and indexes every transaction and then for each successful transaction it performs some action for each of the multi-actions that were included.

File: extensions/indexer/indexer.go
Comment (aaronbuchwald) [reply extensions/indexer/indexer.go:50]: `result.Success` just indicates whether the execution result was successful. Regardless of that status, the transaction was included in the block. Failed transactions can still pay fees and should be indexed.

File: extensions/indexer/indexer.go
Comment (aaronbuchwald) [reply extensions/indexer/indexer.go:24]: Wasn't sure if that should be part of this PR, but can add it

File: extensions/indexer/indexer.go
Comment (aaronbuchwald) [reply extensions/indexer/indexer.go:42]: Ya

File: extensions/indexer/indexer.go
Comment (aaronbuchwald) [reply extensions/indexer/indexer.go:22]: I'll move the actions to their own accepted subscriber as suggested here https://github.com/ava-labs/hypersdk/pull/1143#discussion_r1679659760

Comment (aaronbuchwald) [reply extensions/indexer/indexer.go:22]: Still leaving this extra interface in, so that each controller can pass in its own successful tx handler. The alternative would be for both of them to just implement the full interface. Either is fine imo, lmk if you'd prefer the latter.

File: extensions/indexer/storage.go
Comment (aaronbuchwald) [reply extensions/indexer/storage.go:1]: Removed the static functions

File: extensions/indexer/indexer.go
Comment (aaronbuchwald) [reply extensions/indexer/indexer.go]: I think we should make adding a unit test a separate PR because you currently need a builder in order to construct a stateless block...

We need to clean a lot of this code up to make writing unit tests more reasonable, so I'd prefer to defer this to a separate PR and keep the scope of this one a little bit smaller.

**Comment by aaronbuchwald** [comment]
https://github.com/ava-labs/hypersdk/pull/1143/commits/707dad3ce39f85a2dd58ff53cb5fc103de8fe495 fixes https://github.com/ava-labs/hypersdk/actions/runs/9963000960/job/27528179448?pr=1143#step:5:433

Screenshot 2024-07-16 at 16 07 11

where I did not not invoke the combined subscriber correctly (still called the tx indexer instead)

**Comment by aaronbuchwald** [comment]
> [707dad3](https://github.com/ava-labs/hypersdk/commit/707dad3ce39f85a2dd58ff53cb5fc103de8fe495) fixes https://github.com/ava-labs/hypersdk/actions/runs/9963000960/job/27528179448?pr=1143#step:5:433
> 
> Screenshot 2024-07-16 at 16 07 11
> where I did not not invoke the combined subscriber correctly (still called the tx indexer instead)

Digging into this a bit more, this happens for the TokenVM tests when the `actionIndexer` is a subscriber registered AFTER the `txIndexer`. If I reverse the order, the tests pass reliably locally.

I can also fix this test by inserting a sleep statement between marking the block as accepted and invoking the `tcli`. It seems that this code is racey because `controller.Accepted` is actually called from a goroutine processes accepted blocks asynchronously: https://github.com/ava-labs/hypersdk/blob/a061629d14a57898514a1d3549906b8d91f4276a/vm/resolutions.go#L213.

This race condition is not introduced by this change, but it does appear to happen reliably with this change iff the `actionIndexer` comes second.

I think it's reasonable to swap the order, so that this passes CI and add a TODO calling out the race condition in the integration test.

File: examples/morpheusvm/controller/action_indexer.go
Comment (joshua-kim) [inline examples/morpheusvm/controller/action_indexer.go:16]: Should we call this something else...? Something related metrics or something?

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [inline extension/indexer/tx_indexer.go:108]: nit: `NoOp`

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [inline extension/indexer/tx_indexer.go:106]: nit: export type so caller can use this type if it wants to. Applicable to rest of concrete-types in this package

File: extension/indexer/accepted_subscriber.go
Comment (joshua-kim) [inline extension/indexer/accepted_subscriber.go:20]: Can we call this function + this type `func NewAcceptedSubscribers(...) *AcceptedSubscribers`?

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [inline extension/indexer/tx_indexer.go:24]: nit: combine the two `var` blocks

File: examples/morpheusvm/controller/action_indexer.go
Comment (joshua-kim) [inline examples/morpheusvm/controller/action_indexer.go:1]: nit: I think you meant to name this file whatever we end up naming this struct

File: examples/tokenvm/controller/action_indexer.go
Comment (joshua-kim) [inline examples/tokenvm/controller/action_indexer.go:16]: Similar comment as the corresponding comment in morpheusvm

File: examples/tokenvm/controller/action_indexer.go
Comment (joshua-kim) [inline examples/tokenvm/controller/action_indexer.go:36]: Wondering if the order-book updates should happen in their own subscriber since they are logically decoupled from the metrics

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [inline extension/indexer/tx_indexer.go:34]: Could probably just call this `DiskIndexer`. Feels weird to have the concrete-type just be the name of the interface.

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [inline extension/indexer/tx_indexer.go:30]: Probably should just replace this line w/ an embed for the `AcceptedSubscriber` interface it since that's the interface we're using to get access to accepted blocks

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [inline extension/indexer/tx_indexer.go:31]: Should we return a struct here instead? We can define an `indexer.Tx` struct type or something that the indexer defines for indexed tx data.

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go:106]: I disagree with this. This is intentionally hiding that detail assuming callers choose to create this type and don't need to know the details or change anything about it after the fact.

Open to changing my mind on this though

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go:24]: I generally prefer a separate `var` block for interface assertions, but don't see that used elsewhere in this repo and do see a combined `var` block with another variable used in one place.

Will update.

File: examples/tokenvm/controller/action_indexer.go
Comment (aaronbuchwald) [reply examples/tokenvm/controller/action_indexer.go:36]: I think it's fine to do more than one thing here.

Doing this async (as it's handled prior to this PR as well) does introduce a race condition because the in-memory order book is not updated during block accept. We may want to support a `SyncAcceptedSubscriber` and `AsyncAcceptedSubscriber` to fix that.

File: examples/morpheusvm/controller/action_indexer.go
Comment (aaronbuchwald) [reply examples/morpheusvm/controller/action_indexer.go:16]: acceptedActionMetricsHandler? This is a bit too long imo and I slightly prefer actionHandler in this case unless you have an alternative suggestion.

File: examples/morpheusvm/controller/action_indexer.go
Comment (aaronbuchwald) [reply examples/morpheusvm/controller/action_indexer.go:1]: Renaming

File: examples/tokenvm/controller/action_indexer.go
Comment (aaronbuchwald) [reply examples/tokenvm/controller/action_indexer.go:16]: Renaming

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go:34]: Renamed to `txDBIndexer` and renaming the constructor as well

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [reply extension/indexer/tx_indexer.go:106]: 1. If you don't want to let the caller know about the type for some reason, then you can just return the interface. Returning an unexported struct makes the type awkward to use. You can't embed, wrap, return a non-exported type from a calling package.
2. At the same time what's the disadvantage of exporting the type? Returning a concrete type gives more flexibility to the caller to use it either as the interface or as the concrete type. I think we should let consumers of the package decide how they want to use our api (this way they can use our indexer interface, wrap it with their own, or use the type).

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go:31]: This is an interface, so we can swap a noop version in if tx indexing is disabled as per https://github.com/ava-labs/hypersdk/pull/1143#discussion_r1679556902

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go]: Done

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go]: Done

File: examples/morpheusvm/controller/action_indexer.go
Comment (aaronbuchwald) [reply examples/morpheusvm/controller/action_indexer.go:1]: Done

File: examples/tokenvm/controller/action_indexer.go
Comment (aaronbuchwald) [reply examples/tokenvm/controller/action_indexer.go]: Done

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go]: Done

File: examples/tokenvm/controller/action_indexer.go
Comment (joshua-kim) [reply examples/tokenvm/controller/action_indexer.go]: Yeah I think long term we're gonna try to get rid of the metrics in here and push it up into hypersdk anyways so... not too concerned

File: extension/indexer/accepted_subscriber.go
Comment (joshua-kim) [inline extension/indexer/accepted_subscriber.go:24]: nit: update to `a` since we updated this struct name

File: extension/indexer/successful_tx_subscriber.go
Comment (joshua-kim) [inline extension/indexer/successful_tx_subscriber.go:31]: I think this is how the current code behaves... but just wondering if we should perform this check or leave it to the implementation. We don't have to worry about it for this PR though.

File: extension/indexer/tx_indexer.go
Comment (joshua-kim) [inline extension/indexer/tx_indexer.go:40]: nit: export the struct + rename the receiver

File: extension/indexer/accepted_subscriber.go
Comment (joshua-kim) [inline extension/indexer/accepted_subscriber.go:13]: I don't think we need this context + in the other tx subscriber interface currently... we could leave it in since I think it's reasonable that you would want a context though (say if you were archiving into s3 or something).

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go]: Exported

File: extension/indexer/accepted_subscriber.go
Comment (aaronbuchwald) [reply extension/indexer/accepted_subscriber.go]: Updated

File: extension/indexer/successful_tx_subscriber.go
Comment (aaronbuchwald) [reply extension/indexer/successful_tx_subscriber.go:31]: Discussed here https://github.com/ava-labs/hypersdk/pull/1143#discussion_r1679916304 and I opted to keep this, so that each VM only handles accepted actions

File: extension/indexer/accepted_subscriber.go
Comment (aaronbuchwald) [reply extension/indexer/accepted_subscriber.go:13]: Ya I think this is reasonable to leave in as part of the interface

File: extension/indexer/accepted_subscriber.go
Comment (aaronbuchwald) [reply extension/indexer/accepted_subscriber.go]: Done

File: extension/indexer/tx_indexer.go
Comment (aaronbuchwald) [reply extension/indexer/tx_indexer.go]: Done

**Review by joshua-kim — APPROVED** [review]

**Pull request merged**
